Native type generators cannot be used for user defined types

```go
package check

import (
  "fmt"
  "testing"
  
  "github.com/steffnova/go-check"
  "github.com/steffnova/go-check/constraints"
  "github.com/steffnova/go-check/generator"
)

type i8 int8

func TestUserInt8(t *testing.T) {
    predicate := func(x i8) error {
        // Property should always fail
        return fmt.Errorf("")
    }
    
    check.Check(t, check.Property(predicate, generator.Int8()))
}

```

Executing this test would result in panic:

```
panic: reflect: Call using int8 as type check.i8 [recovered]
        panic: reflect: Call using int8 as type check.i8
```

[Int64](https://github.com/steffnova/go-check/blob/dda683bc5259ba282db421d5fbb5b481271cdf9b/generator/int.go#L28)  generator is returning int64 as a value which couldn't be passed as user defined type:
```go
type userInt64 int64
```
The generated int64 value needs to be converted to it's target type. This can be done using [Convert method](https://pkg.go.dev/reflect#Value.Convert)

--- a/generator/arbitrary.go
+++ b/generator/arbitrary.go
@@ -40,7 +40,7 @@ func (arb Arbitrary) Map(mapper interface{}) Arbitrary {
 			return func() (reflect.Value, shrinker.Shrinker) {
 				val, shrinker := generator()
 				val = reflect.ValueOf(mapper).Call([]reflect.Value{val})[0]
-				return val, shrinker.Map(mapper)
+				return val.Convert(target), shrinker.Map(target, mapper)
 			}, nil
 		}
 	}
@@ -51,8 +51,8 @@ func (arb Arbitrary) Map(mapper interface{}) Arbitrary {
 // type, while output parameter must be bool. Generator returned by new Arbitrary will
 // generate values that satisfy predicate.
 //
-// NOTE: This can highly impact Generator's time to generate arbitrary.Type as it will
-// try to generate target's values unitl predicate is satisfied.
+// NOTE: This can highly impact Generator's time to generate a value as it will try to
+// generate target's values unitl predicate is satisfied.
 func (arb Arbitrary) Filter(predicate interface{}) Arbitrary {
 	return func(target reflect.Type, r Random) (Generator, error) {
 		generate, err := arb(target, r)
@@ -74,7 +74,7 @@ func (arb Arbitrary) Filter(predicate interface{}) Arbitrary {
 				val, _ := generate()
 				outputs := reflect.ValueOf(predicate).Call([]reflect.Value{val})
 				if outputs[0].Bool() {
-					return val, nil
+					return val.Convert(target), nil
 				}
 			}
 		}, nil

--- a/generator/float.go
+++ b/generator/float.go
@@ -37,7 +37,8 @@ func Float64(limits ...constraints.Float64) Arbitrary {
 
 		return func() (reflect.Value, shrinker.Shrinker) {
 			n := r.Float64(constraint.Min, constraint.Max)
-			return reflect.ValueOf(n), shrinker.Float64(n, constraint)
+			val := reflect.ValueOf(n).Convert(target)
+			return val, shrinker.Float64(val, constraint)
 		}, nil
 	}
 }

--- a/generator/int.go
+++ b/generator/int.go
@@ -25,7 +25,8 @@ func Int64(limits ...constraints.Int64) Arbitrary {
 		}
 		return func() (reflect.Value, shrinker.Shrinker) {
 			n := r.Int64(constraint.Min, constraint.Max)
-			return reflect.ValueOf(n), shrinker.Int64(n, constraint)
+			nVal := reflect.ValueOf(n).Convert(target)
+			return nVal, shrinker.Int64(nVal, constraint)
 		}, nil
 	}
 }

--- a/generator/uint.go
+++ b/generator/uint.go
@@ -24,7 +24,8 @@ func Uint64(limits ...constraints.Uint64) Arbitrary {
 		}
 		return func() (reflect.Value, shrinker.Shrinker) {
 			n := r.Uint64(constraint.Min, constraint.Max)
-			return reflect.ValueOf(n), shrinker.Uint64(n, constraint)
+			nVal := reflect.ValueOf(n).Convert(target)
+			return nVal, shrinker.Uint64(nVal, constraint)
 		}, nil
 	}
 }

--- a/property.go
+++ b/property.go
@@ -71,7 +71,12 @@ func Property(predicate interface{}, arbGenerators ...generator.Arbitrary) prope
 				for shrinker != nil {
 					oldValue := inputs[index]
 					failed := !outputs[0].IsZero()
-					inputs[index], shrinker = shrinker(failed)
+					var err error
+					inputs[index], shrinker, err = shrinker(failed)
+					if err != nil {
+						return fmt.Errorf("failed shrink input with index: %d. %s", index, err)
+					}
+
 					if failed && !reflect.DeepEqual(oldValue.Interface(), inputs[index].Interface()) {
 						numberOfShrinks++
 					}

--- a/shrinker/array.go
+++ b/shrinker/array.go
@@ -1,28 +1,37 @@
 package shrinker
 
 import (
+	"fmt"
 	"reflect"
 )
 
 // Array is a shrinker for array. Array is shrinked by shrinking it's elements one at a time
 // Convergence speed for shrinker is O(n*m), n is array size and m is convergance speed of
 // array elements.
-func Array(original reflect.Value, shrinkers []Shrinker) Shrinker {
-	return func(propertyFailed bool) (reflect.Value, Shrinker) {
-		newArray := reflect.New(original.Type()).Elem()
-		reflect.Copy(newArray, original)
+func Array(val reflect.Value, shrinkers []Shrinker) Shrinker {
+	return func(propertyFailed bool) (reflect.Value, Shrinker, error) {
+		if val.Kind() != reflect.Array {
+			return reflect.Value{}, nil, fmt.Errorf("array shrinker cannot shrink %s", val.Kind().String())
+		}
+
+		newArray := reflect.New(val.Type()).Elem()
+		reflect.Copy(newArray, val)
 
 		for index, shrinker := range shrinkers {
 			if shrinker == nil {
 				continue
 			}
-			val, shrinker := shrinker(propertyFailed)
+
+			val, shrinker, err := shrinker(propertyFailed)
+			if err != nil {
+				return reflect.Value{}, nil, fmt.Errorf("failed to shrink array element. %w", err)
+			}
 
 			newArray.Index(index).Set(val)
 			shrinkers[index] = shrinker
-			return newArray, Array(newArray, shrinkers)
+			return newArray, Array(newArray, shrinkers), nil
 		}
 
-		return original, nil
+		return val, nil, nil
 	}
 }

--- a/shrinker/float.go
+++ b/shrinker/float.go
@@ -1,50 +1,62 @@
 package shrinker
 
 import (
+	"fmt"
 	"math"
 	"reflect"
 
 	"github.com/steffnova/go-check/constraints"
 )
 
-// Float64 is a shrinker for float64. N is the shrinking target and limits are constraints
-// in which n will be shrunk. If n > 0 it will be shrunk towards limits.Min or 0, whichever
-// is higher. If n < 0 it will be shrunk towards 0 or limits.Max, whichever is lower. If
-// n == 0 it will be returned with no shrinker as it is converging point in ranges
-// [math.Float64Min, 0] and [0, math.Float64Max]
-func Float64(n float64, limits constraints.Float64) Shrinker {
-	switch {
-	case n > 0:
-		return float64Positive(n, limits)
-	case n < 0:
-		return float64Negative(n, limits)
-	default:
-		return func(propertyFailed bool) (reflect.Value, Shrinker) {
-			return reflect.ValueOf(float64(0)), nil
+// Float64 is a shrinker for float64. Val is the shrinking target and limits are constraints in which
+// val will be shrunk. If val.Float() > 0 it will be shrunk towards limits.Min or 0, whichever is higher.
+// Otherwise it will be shrunk towards 0 or limits.Max, whichever is lower.
+//
+// An error is returned if val's underlying type is not float64, limits.Min is greater than limits.Max
+// or if val's value is out of limit bounds.
+func Float64(val reflect.Value, limits constraints.Float64) Shrinker {
+	return func(propertyFailed bool) (reflect.Value, Shrinker, error) {
+		switch {
+		case val.Kind() != reflect.Float64:
+			return reflect.Value{}, nil, fmt.Errorf("float64 shrinker cannot shrink %s", val.Kind().String())
+		case limits.Min > limits.Max:
+			return reflect.Value{}, nil, fmt.Errorf("limit's min: %f can't be greater than limit's max: %f", limits.Min, limits.Max)
+		case val.Float() < limits.Min || val.Float() > limits.Max:
+			return reflect.Value{}, nil, fmt.Errorf("n: %v is out of limit constraints: {Min: %f, Max: %f}", val.Float(), limits.Min, limits.Max)
+		case val.Float() > 0:
+			return float64Positive(val.Float(), val.Type(), limits)(propertyFailed)
+		default:
+			return float64Negative(-val.Float(), val.Type(), limits)(propertyFailed)
 		}
 	}
 }
 
-// float64Positive is a shrinker for positive float64 numbers. N is the shrinking target and limits are
-// constraints in which n will be shrunk. N will be shrunk towards limits.Min or 0 whichever is higher.
-func float64Positive(n float64, limits constraints.Float64) Shrinker {
-	if limits.Min < 0 {
+// float64Positive is a shrinker for positive float64 numbers. N is the shrinking value, target is a type
+// to which shrunk value will be converted to and limits are constraints in which n will be shrunk . N will
+// be shrunk towards limits.Min or 0 whichever is higher.
+func float64Positive(n float64, target reflect.Type, limits constraints.Float64) Shrinker {
+	if limits.Min <= 0 {
 		limits.Min = 0
 	}
-
-	return Uint64(math.Float64bits(n), constraints.Uint64{
+	return Uint64(reflect.ValueOf(math.Float64bits(n)), constraints.Uint64{
 		Min: math.Float64bits(limits.Min),
 		Max: math.Float64bits(limits.Max),
-	}).Map(func(exponent uint64) float64 {
-		return math.Float64frombits(exponent)
+	}).Map(target, func(floatBits uint64) float64 {
+		return math.Float64frombits(floatBits)
 	})
 }
 
-// float64Negative is a shrinker for negative float64 numbers. N is the shrinking target and limits are
-// constraints in which n will be shrunk. N will be shrunk towards limits.Max or 0 whichever is lower.
-func float64Negative(n float64, limits constraints.Float64) Shrinker {
-	limits.Max, limits.Min = -limits.Min, -limits.Max
-	return float64Positive(-n, limits).Map(func(n float64) float64 {
-		return -n
+// float64Negative is a shrinker for negative float64 numbers. N is the shrinking value, target is a type
+// to which shrunk value will be converted to and limits are constraints in which n will be shrunk . N will
+// be shrunk towards limits.Max or 0 whichever is lower.
+func float64Negative(n float64, target reflect.Type, limits constraints.Float64) Shrinker {
+	if limits.Max >= 0 {
+		limits.Max = math.Copysign(0, -1)
+	}
+	return Uint64(reflect.ValueOf(math.Float64bits(n)), constraints.Uint64{
+		Min: math.Float64bits(-limits.Max),
+		Max: math.Float64bits(-limits.Min),
+	}).Map(target, func(floatBits uint64) float64 {
+		return -math.Float64frombits(floatBits)
 	})
 }

--- a/shrinker/int.go
+++ b/shrinker/int.go
@@ -1,68 +1,59 @@
 package shrinker
 
 import (
+	"fmt"
 	"reflect"
 
 	"github.com/steffnova/go-check/constraints"
 )
 
-// Int64 is a shrinker for int64. N is the shrinking target and limits are
-// constraints in which n will be shrunk. If n > 0 it will be shrunk towards
-// limits.Min or 0 whichever is higher. If n < 0 it will be shrunk towards 0
-// or limits.Max, whichever is lower. If n == 0 it will be returned with no
-// shrinker as it is converging point in ranges [math.Int64Min, 0] and [0, math.Int64Max]
-func Int64(n int64, limits constraints.Int64) Shrinker {
-	switch {
-	case n > 0:
-		return int64Positive(n, limits)
-	case n < 0:
-		return int64Negative(n, limits)
-	default:
-		return func(propertyFailed bool) (reflect.Value, Shrinker) {
-			return reflect.ValueOf(int64(0)), nil
+// Int64 is a shrinker for int64. Val is the shrinking target and limits are constraints in which
+// val will be shrunk. If val.Int() it will be shrunk towards limits.Min or 0, whichever is higher.
+// Otherwise it will be shrunk towards 0 or limits.Max, whichever is lower.
+//
+// An error is returned if val's underlying type is not int64, limits.Min is greater than limits.Max or
+// if val's value is out of limit bounds.
+func Int64(val reflect.Value, limits constraints.Int64) Shrinker {
+	return func(propertyFailed bool) (reflect.Value, Shrinker, error) {
+		switch {
+		case val.Kind() != reflect.Int64:
+			return reflect.Value{}, nil, fmt.Errorf("int64 shrinker cannot shrink %s", val.Kind().String())
+		case limits.Min > limits.Max:
+			return reflect.Value{}, nil, fmt.Errorf("lower limit: %d cannot be greater than upper limit: %d", limits.Min, limits.Max)
+		case val.Int() < limits.Min || val.Int() > limits.Max:
+			return reflect.Value{}, nil, fmt.Errorf("n: %v is out of limit constraints: {Min: %v, Max: %v}", val.Int(), limits.Min, limits.Max)
+		case val.Int() > 0:
+			return int64Positive(val.Int(), val.Type(), limits)(propertyFailed)
+		default:
+			return int64Negative(val.Int(), val.Type(), limits)(propertyFailed)
 		}
 	}
-
 }
 
 // int64Positive is a shrinker of positive int64 numbers. All numbers
 // are shrunk towards 0 or limits.Min whichever is higher.
-func int64Positive(n int64, limits constraints.Int64) Shrinker {
+func int64Positive(n int64, target reflect.Type, limits constraints.Int64) Shrinker {
 	if limits.Min < 0 {
 		limits.Min = 0
 	}
-	return func(propertyFailed bool) (reflect.Value, Shrinker) {
-		switch {
-		case limits.Max == limits.Min:
-			return reflect.ValueOf(n), nil
-		case propertyFailed:
-			limits.Max = n
-		default:
-			limits.Min = n + 1
-		}
-
-		shrinked := limits.Max - ((limits.Max-limits.Min)/2 + (limits.Max-limits.Min)%2)
-		return reflect.ValueOf(shrinked), int64Positive(shrinked, limits)
-	}
+	return Uint64(reflect.ValueOf(uint64(n)), constraints.Uint64{
+		Min: uint64(limits.Min),
+		Max: uint64(limits.Max),
+	}).Map(target, func(n uint64) int64 {
+		return int64(n)
+	})
 }
 
 // int64Negative is a shrinker of negative int64 numbers. All numbers
 // are shrunk towards 0 or limits.Max whichever is lower.
-func int64Negative(n int64, limits constraints.Int64) Shrinker {
+func int64Negative(n int64, target reflect.Type, limits constraints.Int64) Shrinker {
 	if limits.Max > 0 {
 		limits.Max = 0
 	}
-	return func(propertyFailed bool) (reflect.Value, Shrinker) {
-		switch {
-		case limits.Max == limits.Min:
-			return reflect.ValueOf(n), nil
-		case propertyFailed:
-			limits.Min = n
-		default:
-			limits.Max = n - 1
-		}
-
-		shrinked := limits.Max - (limits.Max-limits.Min)/2
-		return reflect.ValueOf(shrinked), int64Negative(shrinked, limits)
-	}
+	return Uint64(reflect.ValueOf(uint64(-n)), constraints.Uint64{
+		Min: uint64(-limits.Max),
+		Max: uint64(-limits.Min),
+	}).Map(target, func(n uint64) int64 {
+		return -int64(n)
+	})
 }

--- a/shrinker/ptr.go
+++ b/shrinker/ptr.go
@@ -1,34 +1,45 @@
 package shrinker
 
-import "reflect"
+import (
+	"fmt"
+	"reflect"
+)
 
-// Ptr is a shrinker for pointers. ElementShrinker is shrinker for type to which
-// ptr points to. Shrinking process consists of shrinking the value to which pointer
-// points to, and shrinking of pointer to nil.
-func Ptr(ptr reflect.Value, elementShrinker Shrinker) Shrinker {
-	var shrinker Shrinker
-	var lastValidPtr reflect.Value
+// Ptr is a shrinker for pointers. Shrinker is shrinker for type to which ptr points to
+// points to. Shrinking process consists of shrinking the pointer's value, and shrinking
+// of pointer to nil
+func Ptr(val reflect.Value, shrinker Shrinker) Shrinker {
+	return func(propertyFailed bool) (reflect.Value, Shrinker, error) {
+		switch {
+		case val.Kind() != reflect.Ptr:
+			return reflect.Value{}, nil, fmt.Errorf("ptr shrinker cannot shrink: %s", val.Kind().String())
+		default:
+			return ptr(val, val, shrinker)(propertyFailed)
+		}
+	}
+}
 
-	shrinker = func(propertyFailed bool) (reflect.Value, Shrinker) {
-		var val reflect.Value
-		if elementShrinker != nil {
-			val, elementShrinker = elementShrinker(propertyFailed)
-			ptr.Elem().Set(val)
-			return ptr, shrinker
+func ptr(val reflect.Value, lastVal reflect.Value, shrinker Shrinker) Shrinker {
+	return func(propertyFailed bool) (reflect.Value, Shrinker, error) {
+		if shrinker != nil {
+			ptrVal, shrinker, err := shrinker(propertyFailed)
+			if err != nil {
+				return reflect.Value{}, nil, fmt.Errorf("failed to shrink ptr's value. %w", err)
+			}
+			val.Elem().Set(ptrVal)
+			return val, ptr(val, val, shrinker), nil
 		}
 
 		switch {
-		case !ptr.IsNil():
+		case !val.IsNil():
 			// Shrinking to nil could make a property not fail. The last valid pointer
 			// value needs to be saved in order to revert nil pointer if needed.
-			lastValidPtr, ptr = ptr, reflect.Zero(ptr.Type())
-			return ptr, shrinker
+			lastVal, val = val, reflect.Zero(val.Type())
+			return val, ptr(val, lastVal, shrinker), nil
 		case propertyFailed:
-			return ptr, nil
+			return val, nil, nil
 		default:
-			return lastValidPtr, nil
+			return lastVal, nil, nil
 		}
 	}
-
-	return shrinker
 }

--- a/shrinker/shrinker.go
+++ b/shrinker/shrinker.go
@@ -1,25 +1,45 @@
 package shrinker
 
-import "reflect"
+import (
+	"fmt"
+	"reflect"
+)
 
 // Shrinker returns shrinked value and next shrinker. Returned values
 // are usually affected by propertyFailed parameter. It helps shrinker
 // to decide how to properly shrink the value. If returned Shrinker
 // is nil, it indicates that value can no longer be shrinked
-type Shrinker func(propertyFailed bool) (reflect.Value, Shrinker)
+type Shrinker func(propertyFailed bool) (reflect.Value, Shrinker, error)
 
-// Map maps the shrinked value to a new one using the mapper. Mapper
-// must be a function with one input and one output parameter. Input
-// parameter must match shrinked type, otherwise panic occurs.
-func (shrinker Shrinker) Map(mapper interface{}) Shrinker {
-	return func(propertyFailed bool) (reflect.Value, Shrinker) {
-		shrink, shrinker := shrinker(propertyFailed)
-		shrink = reflect.ValueOf(mapper).Call([]reflect.Value{shrink})[0]
+// Map maps the shrinked value to a new one using the mapper and target.
+// Mapper must be a function with one input and one output parameter where
+// input parameter must match shrinked type, otherwise panic occurs. Target
+// is used to convert mapped value to a expected type.
+func (shrinker Shrinker) Map(target reflect.Type, mapper interface{}) Shrinker {
+	return func(propertyFailed bool) (reflect.Value, Shrinker, error) {
+		mapperVal := reflect.ValueOf(mapper)
 
-		if shrinker == nil {
-			return shrink, nil
+		switch {
+		case mapperVal.Kind() != reflect.Func:
+			return reflect.Value{}, nil, fmt.Errorf("mapper must be a function")
+		case mapperVal.Type().NumIn() != 1:
+			return reflect.Value{}, nil, fmt.Errorf("mapper must have 1 input value")
+		case mapperVal.Type().NumOut() != 1:
+			return reflect.Value{}, nil, fmt.Errorf("mapper must have 1 output value")
+		case !mapperVal.Type().Out(0).ConvertibleTo(target):
+			return reflect.Value{}, nil, fmt.Errorf("mapper's output: %s can't be converted to target: %s", mapperVal.Type().Out(0).String(), target.String())
 		}
 
-		return shrink, shrinker.Map(mapper)
+		shrink, shrinker, err := shrinker(propertyFailed)
+		if err != nil {
+			return reflect.Value{}, nil, fmt.Errorf("shrinker.Map shrinking failed. %w", err)
+		}
+
+		shrink = mapperVal.Call([]reflect.Value{shrink})[0].Convert(target)
+
+		if shrinker == nil {
+			return shrink, nil, nil
+		}
+		return shrink, shrinker.Map(target, mapper), nil
 	}
 }

--- a/shrinker/slice.go
+++ b/shrinker/slice.go
@@ -1,6 +1,7 @@
 package shrinker
 
 import (
+	"fmt"
 	"reflect"
 
 	"github.com/steffnova/go-check/constraints"
@@ -16,40 +17,44 @@ func Slice(target reflect.Type, original []reflect.Value, elementShrinkers []Shr
 	candidateIndex := 0
 	shrinker := Shrinker(nil)
 
-	shrinker = func(propertyFailed bool) (reflect.Value, Shrinker) {
+	shrinker = func(propertyFailed bool) (reflect.Value, Shrinker, error) {
 		value := reflect.MakeSlice(target, 0, 0)
 		elementsShrunk := true
-
+		var err error
 		for index, elementShrinker := range elementShrinkers {
 			if elementShrinker == nil {
 				continue
 			}
 			elementsShrunk = false
-			original[index], elementShrinkers[index] = elementShrinker(propertyFailed)
+			original[index], elementShrinkers[index], err = elementShrinker(propertyFailed)
+			if err != nil {
+				return reflect.Value{}, nil, fmt.Errorf("failed to shrink slice element at index: %d. %w", index, err)
+			}
+			return reflect.Append(value, original...), Slice(target, original, elementShrinkers, limits), nil
 		}
 
 		switch {
 		case !elementsShrunk:
 			// If elements are not shrunk to smalles possible value, keep shrinking them
-			return reflect.Append(value, original...), shrinker
+			return reflect.Append(value, original...), shrinker, nil
 		case !propertyFailed:
 			// If element removed from a slice causes the property to succeed, it needs to
 			// be added to a candidates as it is essential element for property failure
 			candidates = append(candidates, original[candidateIndex-1])
-			return reflect.Append(value, append(candidates, original[candidateIndex:]...)...), shrinker
+			return reflect.Append(value, append(candidates, original[candidateIndex:]...)...), shrinker, nil
 		case limits.Min == len(original)-(candidateIndex-len(candidates)):
 			// Shrinking should stop if we've reached a slice's minimal length defined by constraint.
 			// The last shrunk value is aggregation of candidates and remaning original elements
-			return reflect.Append(value, append(candidates, original[candidateIndex:]...)...), nil
+			return reflect.Append(value, append(candidates, original[candidateIndex:]...)...), nil, nil
 		case candidateIndex == len(original):
 			// Shrinking should stop if candidate index has passed through all elements.
 			// Candidates is final shrunk value of original slice
-			return reflect.Append(value, candidates...), nil
+			return reflect.Append(value, candidates...), nil, nil
 		default:
 			// In all other cases property keeps failing so slice size shrinking continues
 			// TODO: See if size shrinking speed can be increased
 			candidateIndex++
-			return reflect.Append(value, append(candidates, original[candidateIndex:]...)...), shrinker
+			return reflect.Append(value, append(candidates, original[candidateIndex:]...)...), shrinker, nil
 		}
 	}
 	return shrinker

--- a/shrinker/struct.go
+++ b/shrinker/struct.go
@@ -1,30 +1,38 @@
 package shrinker
 
 import (
+	"fmt"
 	"reflect"
 )
 
 // Struct is a shrinker for Go's struct. Struct shrinking consists of shrinking individual
 // fields one by one. Convergance speed for shrinker is O(n*m), n is number of Struct fields
 // and m is convergance speed of field type.
-func Struct(original reflect.Value, fieldShrinkers map[string]Shrinker) Shrinker {
-	return func(propertyFailed bool) (reflect.Value, Shrinker) {
-		for i := 0; i < original.Type().NumField(); i++ {
-			fieldName := original.Type().Field(i).Name
+func Struct(val reflect.Value, fieldShrinkers map[string]Shrinker) Shrinker {
+	return func(propertyFailed bool) (reflect.Value, Shrinker, error) {
+		if val.Kind() != reflect.Struct {
+			return reflect.Value{}, nil, fmt.Errorf("struct shrinker cannot shrink: %s", val.Kind().String())
+		}
+
+		for i := 0; i < val.Type().NumField(); i++ {
+			fieldName := val.Type().Field(i).Name
 
 			shrinker := fieldShrinkers[fieldName]
 			if shrinker == nil {
 				continue
 			}
 
-			val, shrinker := shrinker(propertyFailed)
+			fieldVal, shrinker, err := shrinker(propertyFailed)
+			if err != nil {
+				return reflect.Value{}, nil, fmt.Errorf("failed to shrink struct field: %s. %w", fieldName, err)
+			}
 
-			original.FieldByName(fieldName).Set(val)
+			val.FieldByName(fieldName).Set(fieldVal)
 			fieldShrinkers[fieldName] = shrinker
 
-			return original, Struct(original, fieldShrinkers)
+			return val, Struct(val, fieldShrinkers), nil
 		}
 
-		return original, nil
+		return val, nil, nil
 	}
 }

--- a/shrinker/uint.go
+++ b/shrinker/uint.go
@@ -1,26 +1,37 @@
 package shrinker
 
 import (
+	"fmt"
 	"reflect"
 
 	"github.com/steffnova/go-check/constraints"
 )
 
-// Uint64 is a shrinker for uint64. N is the shrinking target and limits are
-// constraints in which n will be shrunk. N will be shrunk towards limits.Min
-// or 0 whichever is higher.
-func Uint64(n uint64, limits constraints.Uint64) Shrinker {
-	return func(propertyFailed bool) (reflect.Value, Shrinker) {
+// Uint64 is a shrinker for uint64. Val is the shrinking target and limits are constraints in which val
+// will be shrunk. Val will be shrunk towards limits.Min or 0, whichever is higher.
+//
+// An error is returned if val's underlying type is not uint64, limits.Min is greater than limits.Max
+// or if val's value is out of limit bounds.
+func Uint64(val reflect.Value, limits constraints.Uint64) Shrinker {
+	return func(propertyFailed bool) (reflect.Value, Shrinker, error) {
 		switch {
+		case val.Kind() != reflect.Uint64:
+			return reflect.Value{}, nil, fmt.Errorf("uint64 shrinker cannot shrink %s", val.Kind().String())
+		case limits.Min > limits.Max:
+			return reflect.Value{}, nil, fmt.Errorf("lower limit: %d cannot be greater than upper limit: %d", limits.Min, limits.Max)
+		case val.Uint() < limits.Min || val.Uint() > limits.Max:
+			return reflect.Value{}, nil, fmt.Errorf("n: %v is out of limit constraints: {Min: %v, Max: %v}", val.Uint(), limits.Min, limits.Max)
 		case limits.Max == limits.Min:
-			return reflect.ValueOf(n), nil
+			return val, nil, nil
 		case propertyFailed:
-			limits.Max = n
+			limits.Max = val.Uint()
 		default:
-			limits.Min = n + 1
+			limits.Min = val.Uint() + 1
 		}
 
-		shrinked := limits.Max - ((limits.Max-limits.Min)/2 + (limits.Max-limits.Min)%2)
-		return reflect.ValueOf(shrinked), Uint64(shrinked, limits)
+		shrink := limits.Max - ((limits.Max-limits.Min)/2 + (limits.Max-limits.Min)%2)
+		shrinkVal := reflect.ValueOf(shrink).Convert(val.Type())
+
+		return shrinkVal, Uint64(shrinkVal, limits), nil
 	}
 }

